File: course1_algorithmic_toolbox/week3_greedy_algorithms/6_maximum_number_of_prizes/DifferentSummands.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Scanner;
import java.util.Set;

public class DifferentSummands {

	private static List<Integer> optimalSummands(int n) {
		Set<Integer> summands = new HashSet<>();

		long sum = 0;

		for (int i = 1; i <= n; i++) {
			int left = (int) (n - (sum + i));

			if (left == i || summands.contains(left)) {
				continue;
			}

			if (left < 0) {
				break;
			}

			summands.add(i);
			sum += i;
		}

		return new ArrayList<>(summands);
	}

	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		int n = scanner.nextInt();
		List<Integer> summands = optimalSummands(n);
		System.out.println(summands.size());
		for (Integer summand : summands) {
			System.out.print(summand + " ");
		}
	}
}

File: course1_algorithmic_toolbox/week3_greedy_algorithms/6_maximum_number_of_prizes/different_summands.js
const readline = require('readline');
const rl = readline.createInterface({
	input: process.stdin,
	terminal: false
});

process.stdin.setEncoding('utf8');

const optimalSummands = (n) => {
	const summands = new Set();

	let sum = 0;

	for (let i = 1; i <= n; i++) {
		const left = n - (sum + i);

		if (left === i || summands.has(left)) {
			continue;
		}

		if (left < 0) {
			break;
		}

		summands.add(i);
		sum += i;
	}

	return [...summands];
};

/**
 * Enter:
 *
 * 8
 *
 */

rl.once('line', line => {
	const summands = optimalSummands(parseInt(line, 10));
	console.log(summands.length);
	for (const summand of summands) {
		process.stdout.write(`${summand} `);
	}
	process.exit();

});


module.exports = optimalSummands;
